apps/token-manager/app/src/screens/Holders.js

```diff
@@ -1,166 +1,163 @@
-                        )}
-                        maxAccountTokens={maxAccountTokens}
-                        tokenDecimalsBase={tokenDecimalsBase}
-                        onAssignTokens={onAssignTokens}
-                        onRemoveTokens={onRemoveTokens}
-                        compact={compactTable}
-                      />
-                    ))}
-                  </ResponsiveTable>
-                )}
-              </Main>
-              {(!tabbedNavigation || selectedTab === 1) && (
-                <SideBar
-                  holders={holders}
-                  tokenAddress={tokenAddress}
-                  tokenDecimalsBase={tokenDecimalsBase}
-                  tokenName={tokenName}
-                  tokenSupply={tokenSupply}
-                  tokenSymbol={tokenSymbol}
-                  tokenTransfersEnabled={tokenTransfersEnabled}
-                  userAccount={userAccount}
-                />
-              )}
-            </TwoPanels>
-          )
-        }}
-      </Viewport>
-    )
-  }
+            return values
+          }}
+          renderEntryActions={([address, balance]) => (
+            <EntryActions
+              address={address}
+              balance={balance}
+              onAssignTokens={onAssignTokens}
+              onRemoveTokens={onRemoveTokens}
+              singleToken={balance.eq(tokenDecimalsBase)}
+              canAssign={balance.lt(maxAccountTokens)}
+            />
+          )}
+        />
+      }
+      secondary={
+        <InfoBoxes
+          holders={holders}
+          tokenAddress={tokenAddress}
+          tokenDecimalsBase={tokenDecimalsBase}
+          tokenName={tokenName}
+          tokenSupply={tokenSupply}
+          tokenSymbol={tokenSymbol}
+          tokenTransfersEnabled={tokenTransfersEnabled}
+          userAccount={userAccount}
+        />
+      }
+    />
+  )
+}
 
-  handleSelectTab = index => {
-    this.setState({ selectedTab: index })
-  }
+Holders.propTypes = {
+  holders: PropTypes.array,
+}
+Holders.defaultProps = {
+  holders: [],
 }
 
-const TabBarWrapper = styled.div`
-  margin-top: 16px;
-  & ul {
-    border-bottom: none !important;
-  }
-  & li {
-    padding: 0 20px;
-  }
-`
+function EntryActions({
+  address,
+  onAssignTokens,
+  onRemoveTokens,
+  singleToken,
+  canAssign,
+}) {
+  const theme = useTheme()
+  const [label, showLocalIdentityModal] = useIdentity(address)
 
-const ResponsiveTable = styled(Table)`
-  margin-top: 16px;
+  const editLabel = useCallback(() => showLocalIdentityModal(address), [
+    address,
+    showLocalIdentityModal,
+  ])
+  const assignTokens = useCallback(() => onAssignTokens(address), [
+    address,
+    onAssignTokens,
+  ])
+  const removeTokens = useCallback(() => onRemoveTokens(address), [
+    address,
+    onRemoveTokens,
+  ])
 
-  ${breakpoint(
-    'medium',
-    `
-      opacity: 1;
-      margin-top: 0;
-    `
-  )};
-`
+  const actions = [
```

Comment:
Super nitpick here: 
```suggestion
  const actions = [
    ...[canAssign ? [assignTokens, IconAdd, 'Add tokens'] : []]
```

Using `unshift` below is mutating the contents, nothing big, but if we allow it here, what stops us to do other operations on it elsewhere (or maybe some other part of the code) and I try to avoid it by setting everything in one go.

You can disregard if this is not a concern for you, I just wanted to let you know my line of thinking.

Reply:
What’s not to love about some well contained mutations? :laughing: I know this can be a controversial subject, but I think that deciding to use local immutability, like imperative vs. functional code, should also take into consideration readability and performances.

Having pure functions and components is a big advantage, but trying to achieve immutability everywhere often comes at a readability and performance cost with JavaScript, so I don’t really have a problem with variable mutations as long as it doesn’t have any side effects outside of the current scope. On this regard, I like how Immer is fully embracing this idea: https://medium.com/hackernoon/introducing-immer-immutability-the-easy-way-9d73d8f71cb3

Now about this specific case: I find them both readable, and the performance cost is not relevant here, so let’s go for your suggestion :+1: